### ISSUE TITLE
Can’t find layout template in v9.26.0 (using Eleventy v1.0.0-canary.44)

### ISSUE THREAD

**Issue by paulrobertlloyd** [harttle/liquidjs#395]
Updating the dependencies for my Eleventy-based website (currently using v1.0.0-canary.44), and updating LiquidJS from v9.25.1 to v9.26.0 generates the following error:

```
[11ty] Problem writing Eleventy templates: (more in DEBUG output)
[11ty] > Having trouble writing template: www/bookmarks/2010/06/star_wars_ipad/index.html

`TemplateWriterWriteError` was thrown
[11ty] > ENOENT: Failed to lookup "default" in "src/includes", file:./src/layouts/bookmark.liquid, line:1, col:1

`RenderError` was thrown
[11ty] > ENOENT: Failed to lookup "default" in "src/includes"

`Error` was thrown:
[11ty]     Error: ENOENT: Failed to lookup "default" in "src/includes"
        at Liquid.lookupError (/Users/paulrobertlloyd/Sites/paulrobertlloyd-v4/node_modules/liquidjs/dist/liquid.node.cjs.js:3121:21)
        at Liquid._parseFile (/Users/paulrobertlloyd/Sites/paulrobertlloyd-v4/node_modules/liquidjs/dist/liquid.node.cjs.js:3063:20)
        at _parseFile.next (<anonymous>)
        at reduce (/Users/paulrobertlloyd/Sites/paulrobertlloyd-v4/node_modules/liquidjs/dist/liquid.node.cjs.js:1737:25)
        at async TemplateLayout.render (/Users/paulrobertlloyd/Sites/paulrobertlloyd-v4/node_modules/@11ty/eleventy/src/TemplateLayout.js:160:25)
        at async Template.renderPageEntry (/Users/paulrobertlloyd/Sites/paulrobertlloyd-v4/node_modules/@11ty/eleventy/src/Template.js:781:17)
        at async /Users/paulrobertlloyd/Sites/paulrobertlloyd-v4/node_modules/@11ty/eleventy/src/Template.js:803:21
        at async Promise.all (index 0)
        at async Promise.all (index 9)
        at async Eleventy.executeBuild (/Users/paulrobertlloyd/Sites/paulrobertlloyd-v4/node_modules/@11ty/eleventy/src/Eleventy.js:958:13)
``` 

As to some background, I have Eleventy set up to look for includes and layouts in separate specified folders. At the top of my templates, I extend a common `default.liquid` template, i.e.:

```liquid
{%- layout 'default' -%}
{%- block 'main' -%}
  …
{%- endblock -%}
```

Eleventy looks for this default template in `./src/layouts/`, but since updating to the latest version of LiquidJS, it appears to now be looking for it in `/src/includes/`. Reverting to v9.25.1 of LiquidJS prevents this error from occurring. Looks like there was a bit of refactoring in the most recent release, which may be the culprit!

Happy to provide more details if required. My site’s code can be found at https://github.com/paulrobertlloyd/paulrobertlloyd-v4

**Comment by harttle** [comment]
Seems 11ty is relying on `.render(ctx, scope, **opts**)`. I removed it (only set when constructing) to be more static and performanc. I'll try to see what impact on perf if I get it back.

**Comment by harttle** [comment]
Supporting customize LiquidOptions on `.render()` have 10%-50% performance impact (cache enabled, since we care about production perf only). After some investigation, the bottleneck is customizing LiquidOptions will postpone cache validation. For example:

* options is static (`root`, `cache`, etc.): by `{% render "foo" %}`, we can directly check cache with key `"foo"`.
* options can be determined during: we need resolve `"foo"` from filesystem against currrent `root`, then check cache with key like `/www/includes/foo.liquid`.

I think we can do this differently to allow @paulrobertlloyd 's scenario and achieve good perf at the same time. Maybe:

* let 11ty have different liquid engine instances for different root configs.
* support root for `includes` and `layout` natively in LiquidJS

The second approach maybe better, but still need LiquidJS make changes and 11ty then make corresponding changes (move these root configs from `render` to `new Liquid`). And I want to remove `opts` from `.render()` for good, not sure whether there's elsewhere 11ty depends on it. @zachleat how do you think?

**Comment by harttle** [comment]
Maybe we take the second approach and @paulrobertlloyd can manually specify these roots. Is providing 2 more constructor parameters `partialsRoot` and `layoutsRoot` sufficient for you? @paulrobertlloyd

**Comment by paulrobertlloyd** [comment]
Sounds good to me! I assume this is something 11ty would do, taking the values for `dir.layouts` and `dir.includes` and passing them on to Liquid?

**Comment by harttle** [comment]
Yes, I believe so. I mean before 11ty can manage to change accordingly, can you use this LiquidJS feature directly so this regression unblocks your work?

**Comment by paulrobertlloyd** [comment]
Oooh, I understand. I can use my own Liquid instance with 11ty again and use these new options, certainly. Might me doing so be a good way to test this is a good approach, as well?

**Comment by paulrobertlloyd** [comment]
Hmm, not sure if this is working. With the following options:

```js
new Liquid({
  extname: '.liquid',
  layouts: './src/layouts',
  partials: './src/includes',
});
```

I get the following error:

```bash
[11ty] Problem writing Eleventy templates: (more in DEBUG output)
[11ty] > Having trouble rendering liquid template ./src/content/articles/1999-03-01-early_graphic_design_in_television.md

`TemplateContentRenderError` was thrown
[11ty] > ENOENT: Failed to lookup "photos" in ".", file:./src/content/articles/1999-03-01-early_graphic_design_in_television.md, line:44, col:1

`RenderError` was thrown
[11ty] > ENOENT: Failed to lookup "photos" in "."

`Error` was thrown:
[11ty]     Error: ENOENT: Failed to lookup "photos" in "."
        at Loader.lookupError (/Users/paulrobertlloyd/Sites/paulrobertlloyd-v4/node_modules/liquidjs/dist/liquid.node.cjs.js:847:21)
        at Loader.lookup (/Users/paulrobertlloyd/Sites/paulrobertlloyd-v4/node_modules/liquidjs/dist/liquid.node.cjs.js:832:20)
        at lookup.next (<anonymous>)
        at reduce (/Users/paulrobertlloyd/Sites/paulrobertlloyd-v4/node_modules/liquidjs/dist/liquid.node.cjs.js:913:25)
```

However, I get no errors if I set an array of directories for `root` instead:

```js
new Liquid({
  extname: '.liquid',
  root: ['./src/layouts', './src/includes'],
});
```

**Comment by zachleat** [comment]
Canary 45 will use liquid 9.27.x—I’ll push this out today

**Comment by zachleat** [comment]
Oh, hmm—maybe it will, looks like we had some liquidjs regressions in our test suite

**Comment by zachleat** [comment]
Eleventy specifically only uses the compile time options to pass in the input file path to support relative includes e.g. `include ../up-one-directory.liquid`.

Is there a better way to do this?

**Comment by zachleat** [comment]
Just as info, it looks like 9.26 introduced regressions specifically with the issues documented above which are being picked up by all of the 1.0 canaries (we tied to `^9.25.1` in canary 44, which resolves to 9.27.0). I think I’ll need to lock to 9.25.x until we iron those out

**Comment by harttle** [comment]
It works on integration tests, not sure where's the difference. Just added a demo for `partials` and `layouts` here: https://github.com/harttle/liquidjs/blob/92bfc65e0b1d937c00a8368b272223c702132d23/demo/nodejs/index.js#L7-L11

@paulrobertlloyd  BTW, the above error `Failed to lookup "photos" in "."` is incorrect in the `"."` part, that's still using `root`. It's fixed on 9.27.1

> uses the compile time options to pass in the input file path to support relative includes e.g. include ../up-one-directory.liquid.

@zachleat can be done by setting to `.partials` and `.layouts` instead of `.root` here: https://github.com/11ty/eleventy/blob/2824db33bab6ef5e8790f2f15a651d331c146f30/src/Engines/Liquid.js#L232-L235

Maybe it's cleaner to be implemented in LiquidJS, what do you think?

**Comment by paulrobertlloyd** [comment]
This is still not working with v9.27.1… I get the same error. Do I need to set `root` as well as `partials` and `layout`? (I tried that, and using a variety of combinations, but only setting an array of `root`s appears to work).

**Comment by paulrobertlloyd** [comment]
> I guess there's something with 11ty, maybe it's setting `root` again in `.compile()`.

Ah, maybe changing settings even in a custom instance don’t affect how 11ty then handles LiquidJS internally. I’ll wait for the downstream update 🙂

**Comment by harttle** [comment]
Add support on v9.28.0 for [relative reference](https://liquidjs.com/tutorials/options.html#Partials-Layouts). Things like following are now natively supported:

```liquid
{% render ../foo/bar %}
{% layout ./bar %}
```

Note that relatively referenced files are also need to be within corresponding root.

**Comment by paulrobertlloyd** [comment]
Can conform that using `layouts` and `partials` options in place of `root` now works as of v9.28.0 🎉

**Comment by zachleat** [comment]
I appreciate the movement here but I do think this was a semver breaking change, right?

Are we not allowed to use `root` option in the third argument to `render` any more?

Should this have been bundled with a major version API change?

**Comment by harttle** [comment]
Added 2 test cases for demo: https://github.com/harttle/liquidjs/blob/1de6adf308cba3bb088244794bc9fb1585eab5a5/test/integration/liquid/liquid.ts#L117-L138

**Comment by zachleat** [comment]
> I made a mistake thinking that argument is only used by myself in .express(). I'll stick to semver next time :)

Just looking at the docs, I think I was a bit harsh! I don’t think you put the third argument there into your API docs—so I’m not sure how I found it. Thank you for the help resolving this though!

**Comment by zachleat** [comment]
Just retested in 9.28.2 and still have a bunch of very basic regressions with `includes` 👀

**Comment by jaredwray** [comment]
I am seeing the same issues with regressions on https://github.com/jaredwray/ecto anyway I can help?

**Comment by jaredwray** [comment]
I am also seeing issues with render and include still. Any status on this?

**Comment by zachleat** [comment]
The fix for this has made it into Eleventy master and will ship with 1.0.0-beta.5.

Tracking at https://github.com/11ty/eleventy/issues/1995
